File: app/Models/Post.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Support\Facades\Storage;

class Post extends Model
{
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'title',
        'slug',
        'excerpt',
        'body',
        'featured_image',
        'is_published',
        'published_at',
        'author_id',
    ];

    protected $casts = [
        'is_published' => 'boolean',
        'published_at' => 'datetime',
    ];

    public function author(): BelongsTo
    {
        return $this->belongsTo(User::class, 'author_id');
    }

    public function scopePublished($query)
    {
        return $query->where('is_published', true)
            ->whereNotNull('published_at')
            ->where('published_at', '<=', now());
    }

    public function getFeaturedImageUrlAttribute(): ?string
    {
        if (!$this->featured_image) {
            return null;
        }
        // sudah S3 key (mengandung slash) -> buat URL
        if (str_contains($this->featured_image, '/')) {
            return Storage::disk('s3')->url($this->featured_image);
        }
        // legacy: URL eksternal lama
        return $this->featured_image;
    }
}

File: logikraf-v2/internal/domain/model/post.go
package model

import "time"

type Post struct {
	ID            uint       `gorm:"primaryKey" json:"id"`
	Title         string     `gorm:"size:255;not null" json:"title"`
	Slug          string     `gorm:"size:255;uniqueIndex;not null" json:"slug"`
	Excerpt       string     `gorm:"type:text" json:"excerpt"`
	Body          string     `gorm:"type:text" json:"body"`
	FeaturedImage string     `gorm:"size:500" json:"featured_image"`
	IsPublished   bool       `gorm:"default:false" json:"is_published"`
	PublishedAt   *time.Time `json:"published_at"`
	AuthorID      uint       `json:"author_id"`
	CreatedAt     time.Time  `json:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at"`
	DeletedAt     *time.Time `gorm:"index" json:"-"`
}
